cmd/kops/toolbox_bundle.go

```diff
@@ -0,0 +1,215 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/golang/glog"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kops/cmd/kops/util"
+	"k8s.io/kops/pkg/bundle"
+	"k8s.io/kops/upup/pkg/kutil"
+	"k8s.io/kops/util/pkg/vfs"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	"k8s.io/kubernetes/pkg/kubectl/util/i18n"
+)
+
+var (
+	toolbox_bundle_long = templates.LongDesc(i18n.T(`
+	Creates a bundle for enrolling a bare metal machine.`))
+
+	toolbox_bundle_example = templates.Examples(i18n.T(`
+	# Bundle
+	kops toolbox bundle --name k8s-cluster.example.com
+	`))
+
+	toolbox_bundle_short = i18n.T(`Bundle cluster information`)
+)
+
+type ToolboxBundleOptions struct {
+	// Target is the machine we are enrolling in the cluster
+	Target string
+}
+
+func (o *ToolboxBundleOptions) InitDefaults() {
+}
+
+func NewCmdToolboxBundle(f *util.Factory, out io.Writer) *cobra.Command {
+	options := &ToolboxBundleOptions{}
+	options.InitDefaults()
+
+	cmd := &cobra.Command{
+		Use:     "bundle",
+		Short:   toolbox_bundle_short,
+		Long:    toolbox_bundle_long,
+		Example: toolbox_bundle_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			err := RunToolboxBundle(f, out, options, args)
+			if err != nil {
+				exitWithError(err)
+			}
+		},
+	}
+
+	cmd.Flags().StringVar(&options.Target, "target", options.Target, "machine to target (IP address)")
+
+	return cmd
+}
+
+func RunToolboxBundle(context Factory, out io.Writer, options *ToolboxBundleOptions, args []string) error {
+	if len(args) == 0 {
+		return fmt.Errorf("Specify name of instance group for node")
+	}
+	if len(args) != 1 {
+		return fmt.Errorf("Can only specify one instance group")
+	}
+
+	if options.Target == "" {
+		return fmt.Errorf("target is required")
+	}
+	groupName := args[0]
+
+	cluster, err := rootCommand.Cluster()
+	if err != nil {
+		return err
+	}
+
+	clientset, err := context.Clientset()
+	if err != nil {
+		return err
+	}
+
+	ig, err := clientset.InstanceGroupsFor(cluster).Get(groupName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error reading InstanceGroup %q: %v", groupName, err)
+	}
+	if ig == nil {
+		return fmt.Errorf("InstanceGroup %q not found", groupName)
+	}
+
+	builder := bundle.Builder{
+		Clientset: clientset,
+	}
+	bundleData, err := builder.Build(cluster, ig)
+	if err != nil {
+		return fmt.Errorf("error building bundle: %v", err)
+	}
+
+	sshUser := os.Getenv("USER")
```

Comment:
I would rather have this as part of the cli options

Reply:
The user is rarely going to be clove, but will be root-foocampany, or root itself. Thoughts?

Reply:
Agreed & good point.  This toolbox is really a building block for the machine controller though, coming up next :-)